File: leetcode/95-UniqueBinarySearchTreesII/uniqueBinarySearchTreesII.py
# Given an integer n, generate all structurally unique BST's (binary search trees) that store values 1...n.
#
# For example,
# Given n = 3, your program should return all 5 unique BST's shown below.
#
#    1         3     3      2      1
#     \       /     /      / \      \
#      3     2     1      1   3      2
#     /     /       \                 \
#    2     1         2                 3

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def list2Tree(self, nums):
        for num in nums:
            self.root = self.insert(num, self.root)
        return self.root

    def insert(self, x, root):
        if not root:
            root = TreeNode(x)
        else:
            if x < root.val:
                root.left = self.insert(x, root.left)
            elif x > root.val:
                root.right = self.insert(x, root.right)
            # if x is in the tree already, we do nothing
        return root

    def printTree(self, root):
        def bst_print_dot_aux(root):
            if root:
                if root.left:
                    f.write("%d->%d;\n" % (root.val, root.left.val))
                    bst_print_dot_aux(root.left)
                if root.right:
                    f.write("%d->%d;\n" % (root.val, root.right.val))
                    bst_print_dot_aux(root.right)
        f = open("bst.dot", "w")
        f.write("digraph bst {\n")
        bst_print_dot_aux(root)
        f.write("}\n")
        f.close()

    def printTree2(self, root):
        # print the tree in the a list format that is the same as the leetcode's
        # 1
        #  \
        #   2      ->  ['1', 'null', '2', 'null', '3']
        #    \
        #     3
        res = []
        def printTree2_helper(res, root):
            if root:
                res.append(str(root.val))
                if root.left or root.right:
                    printTree2_helper(res, root.left)
                    printTree2_helper(res, root.right)
            else:
                res.append("null")
        printTree2_helper(res, root)
        return res

class Solution:
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        def genTrees(start, end):
            list = []
            if start > end:
                list.append(None)
            for idx in range(start, end+1):
                leftList = genTrees(start, idx-1)
                rightList = genTrees(idx+1, end)
                for left_root in leftList:
                    for right_root in rightList:
                        root = TreeNode(idx)
                        root.left = left_root
                        root.right = right_root
                        list.append(root)
            return list
        if n == 0:
            return []
        return genTrees(1, n)


if __name__ == "__main__":
    bst = BST()
    # root = bst.list2Tree([1,2,3])
    # bst.printTree(root)
    # print(bst.printTree2(root))

    sol = Solution()
    list = sol.generateTrees(3)
    for root in list:
        print(bst.printTree2(root))

File: leetcode/95-UniqueBinarySearchTreesII/uniqueBinarySearchTreesII.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <vector>
#include "cppinclude/bt.h"

using namespace std;

class Solution
{
public:
    vector<TreeNode *>
    generateTrees(int n)
    {
        if (n == 0)
        {
            return {};
        }
        return generateTreesHelper(1, n);
    }
private:
    vector<TreeNode*> generateTreesHelper(int start, int end)
    {
        vector<TreeNode*> res;
        if (start > end)
        {
            res.push_back(nullptr);
        }
        for (int i = start; i < end + 1; ++i)
        {
            auto left_roots = generateTreesHelper(start, i-1);
            auto right_roots = generateTreesHelper(i+1, end);
            for (auto & left_root : left_roots)
            {
                for (auto& right_root: right_roots)
                {
                    TreeNode* root = new TreeNode(i);
                    root->left = left_root;
                    root->right = right_root;
                    res.push_back(root);
                }
            }
        }
        return res;
    }
};

using ptr2generateTrees = vector<TreeNode *> (Solution::*)(int);

void test(ptr2generateTrees pfcn)
{
    Solution sol;
    // TODO: Add assertions
    (sol.*pfcn)(3);
}

int main()
{
    ptr2generateTrees pfcn = &Solution::generateTrees;
    test(pfcn);
}